### PR TITLE
Filter users and groups a user can share an app with based on permissions

### LINKED ISSUE #268: [ENH] - Filter users and groups a user can share an app with based on JupyterHub permissions
### Feature description

After https://github.com/nebari-dev/jhub-apps/pull/224 is merged, we'll have the ability to share apps with a set of users and groups. 

We need to filter the users and groups in the UI based on the permissions user has, as in a user may only have permission to share with only a certain set of users and groups.

### Value and/or benefit

When sharing an app the users and groups that are displayed in the UI are basically all users and groups, If a user tries to share the app with someone they don't have permission to share with, the sharing will fail silently.

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by aktech**
## Reference Issues or PRs
Fixes #268 

## What does this implement/fix?

_Put a `x` in the boxes that apply_

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds a feature)
- [ ] Breaking change (fix or feature that would cause existing features not to work as expected)
- [ ] Documentation Update
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build related changes
- [ ] Other (please describe):

## Testing

- [x] Did you test the pull request locally?
- [x] Did you add new tests?

#### Text styling

- [ ] The content is written with [plain language](https://www.plainlanguage.gov/guidelines/) (where relevant).
- [ ] If there are headers, they use the proper header tags (with only one level-one header: `H1` or `#` in markdown).
- [ ] All links describe where they link to (for example, check the [Nebari website](https://nebari.dev/)).
- [ ] This content adheres to the Nebari style guides.

#### Non-text content

- [ ] All content is represented as text (for example, images need alt text, and videos need captions or descriptive transcripts).
- [ ] If there are emojis, there are not more than three in a row.
- [ ] Don't use [flashing GIFs or videos](https://www.w3.org/TR/UNDERSTANDING-WCAG20/seizure-does-not-violate.html).
- [ ] If the content were to be read as plain text, it still makes sense, and no information is missing.

### CODE REVIEW

File: jhub_apps/hub_client/hub_client.py [hunk 1]
```diff
@@ -298,17 +301,36 @@ def get_groups(self):
...
+        )
     }
+
+
+def filter_entity_based_on_scopes(scopes, entities, entity_key="user"):
+    # only available in JupyterHub>=5
+    from jupyterhub.scopes import has_scope, expand_scopes
+    allowed_entities_to_read = set()
+    for entity in entities:
+        if has_scope(f'read:{entity_key}s:name!{entity_key}={entity}', expand_scopes(scopes)):
```
Comment (krassowski) [inline jhub_apps/hub_client/hub_client.py:334]: Can `expand_scopes(scopes)` be assigned to a variable and moves out of the loop?

File: jhub_apps/hub_client/hub_client.py [hunk 2]
```diff
@@ -298,17 +301,36 @@ def get_groups(self):
         r.raise_for_status()
         return r.json()
 
+    @requires_user_token
+    def get_user_scopes(self):
+        assert self.token_json
+        assert self.token_json.get("scopes")
```
Comment (krassowski) [inline jhub_apps/hub_client/hub_client.py:307]: Is it intended to raise when scopes are not in the dictionary, or if scopes are empty?

**Review by krassowski — COMMENTED** [review]
Looks good, just two questions

File: jhub_apps/hub_client/hub_client.py (hunk 1 above)
Comment (aktech) [reply jhub_apps/hub_client/hub_client.py]: Yes definitely, my bad

File: jhub_apps/hub_client/hub_client.py (hunk 2 above)
Comment (aktech) [reply jhub_apps/hub_client/hub_client.py:307]: >Is it intended to raise when scopes are not in the dictionary, or if scopes are empty?

This is intended for when scopes are not in dictionary, this is for when if jupyterhub API changes or unauthorized requests

Comment (krassowski) [reply jhub_apps/hub_client/hub_client.py:307]: Then would it be better to test for:

```
assert "scopes" in self.token_json
```

Though maybe a user without any scopes is a bit pathological edge case?

Comment (aktech) [reply jhub_apps/hub_client/hub_client.py:307]: > Though maybe a user without any scopes is a bit pathological edge case?

Yep, it's very unlikely unless API changes.

**Review by krassowski — APPROVED** [review]
LGTM, thank you @aktech!
